pkg/virt-operator/resource/apply/reconcile.go

```diff
@@ -57,6 +57,13 @@ import (
 const Duration7d = time.Hour * 24 * 7
 const Duration1d = time.Hour * 24
 
+type DefaultInfraComponentsNodePlacementOption int
```

Comment:
Can you change this to `InfraComponentsNodePlacementOption`? It's a bit unusual to have the word `Default` in a type name.

Reply:
Yeah this was the original name.
However I wanted to emphasize that we're setting a default option here, as if the node-placement is already set (even without node affinity) we won't mutate it. I think that setting a `InfraComponentsNodePlacementOption` to require CP nodes, then get pods that are not requiring CP nodes, is confusing. But setting a `DefaultInfraComponentsNodePlacementOption` might be clearer.

WDYT?

Reply:
I'd prefer `InfraComponentsNodePlacementOption`, Default in a type name is a bit confusing as a value can be the default or not for a type but the type itself cannot be default or not(?).

The fact that one of the options is supplied by default is anyway pretty clear from the code, or you can make it even clearer with a comment just before defining the default value.

Reply:
You can also alias the default value to make it clearer.

Reply:
> I'd prefer InfraComponentsNodePlacementOption, Default in a type name is a bit confusing as a value can be the default or not for a type but the type itself cannot be default or not(?).

A default name in a type is not confusing IMO if the type itself represents a default value. Quite the opposite - it reflects the purpose of the type. A value that is of type `Default..` is a default value for something else, obviously, not for the type itself. I just kind of miss why it is confusing.

BTW, types are being prefixed with default pretty often in k8s, but also in Kubevirt:
```bash
> grep -E -i "^type default" ./kubernetes -R --include=*.go --exclude-dir=vendor | wc -l
54

> grep -E -i "^type default" ./kubevirt -R --include=*.go --exclude-dir=vendor | wc -l
5
```

> The fact that one of the options is supplied by default is anyway pretty clear from the code, or you can make it even clearer with a comment just before defining the default value.

I strongly disagree with this approach.
In order to reason about a function and its parameters you only need to look at the function signature. Having to look at the code is a huge code smell, and having to document such a code with comments is even worse IMHO.

Reply:
I tend to agree with @acardace honestly.
If we want to emphasize that it acts as default, it should be the function parameter name that has the `default` (not the type).
Also, what will happen if we decide in the future that this will not be a default? Should we change the type name? Just my cents :)

Reply:
IOW, how about use `InfraComponentsNodePlacementOption` and change the parameter name of `InjectPlacementMetadata` to `defaultNodePlacementOption`?

Reply:
> IOW, how about using `InfraComponentsNodePlacementOption` and changing the parameter name of `InjectPlacementMetadata` to `defaultNodePlacementOption`?

+1

Reply:
> Both function signatures with or without the default in the type clearly say how the parameter will be used.

Can you explain? How would the signature reflect how the parameter will be used without the default?

> If we want to emphasize that it acts as default, it should be the function parameter name that has the default (not the type).

Why it shouldn't be part of the type name if the whole purpose of the type is to represent a default value? If we'd only change the parameter name we make this visible only to the function signature, not the type as a whole.

I'm sorry but I'm really not convinced :) and find it hard to understand your reasoning for why it is confusing. I mean, you claim that it's confusing but don't really explain why. The reasoning for why it is valuable is very simple - it reflects that the type's values serve as default values only. I still can't find a good reason for not reflecting this information, or reflecting it only at the parameter name level.

> Anyway, I think we agree to disagree :)

Sure :)

Reply:
FYI, changed `DefaultInfraComponentsNodePlacementOption` to `DefaultInfraComponentsNodePlacement` (without the option in the end) to make it shorter and more readable.

Reply:
Sorry @fossedihelm, I just now realized I haven't responded to one of your arguments.

> Also, what will happen if we decide in the future that this will not be a default?

What do you mean we'll decide it won't be the default? In that case the type can be deleted, as it's no longer necessary, since its whole purpose is to define a default node placement. Can you elaborate on this one? Which other use we can make of this type in your opinion?

> Should we change the type name?

I don't see why we would ever change it as explained above.

But if there's a hypothetical need to change the name of a completely internal type, are you trying to imply that it's a bad thing to do? Can you explain why or what your concerns are? IMHO doing so is completely valid and again I'm not sure exactly what's the concern / risk / problem and the reasoning for it.

Reply:
>Why shouldn't it be part of the type name if the whole purpose of the type is to represent a default value? If we'd only change the parameter name we make this visible only to the function signature, not the type as a whole.

I think that is here where we disagree. The purpose of the type is to represent the kind of a particular struct.
It is in the `InjectPlacementMetadata` that it takes the form of default value. Outside of the function scope, it is not a default value. You said that it represents a default value, but, thinking outside of the `InjectPlacementMetadata`, the default value of what?
IMO, it becomes a default value when used inside the `InjectPlacementMetadata`, and only there; for this reason to me it is better to change the parameter name.
The same will happen for every structs. We have vms, pods, etc... These types can be used as default for something in some circumstances. But we will change the parameter name, not the type because a pod will always be a pod.
This is my reasoning :) 

>But if there's a hypothetical need to change the name of a completely internal type, are you trying to imply that it's a bad thing to do? Can you explain why or what your concerns are? IMHO doing so is completely valid and again I'm not sure exactly what's the concern / risk / problem and the reasoning for it.

Again, a type should describe the type(kind or whatever), not the behavior :)

Maybe we just have different opinions about it.

Reply:
> Can you explain? How would the signature reflect how the parameter will be used without the default?

`DefaultInfraComponentsNodePlacementOption` vs `InfraComponentsNodePlacementOption`, I just understand a param of this type will be used for the node placement of infra components, `Default` in the type doesn't convey anything, what should default mean? I'm passing an explicit value, how does `Default` tell me how the function will behave? My point is, it's not clear one way or the other.

We're full of examples in our APIs where default values are documented with comments and this is because there's no such thing as a default type in a statically compiled language there can only be default values, but those depend on how the function is used and the value semantics. A default type by itself is a contradiction.

I think what you want to mean is that this value will be used by default if no explicit `componentConfig` is passed in, but the `Default` in the type doesn't tell me that, when I read a type name it must tell me something about the type itself not about how some function using that type should behave when another param is nil.

I think the function is ill-defined, why pass both `componentConfig` and `nodePlacementOption` explicitly? They refer to the same thing, there's a cross dependency that is not clear whether you put the `Default` in the type name or not.  One of the issue is that this function is both used for deployments and daemonsets, and this is why you need the additional parameter, otherwise you wouldn't, you could just encode in the function that if `componentConfig` then the default behavior is `X` or `Y`.

Anyway as I said, even if I don't agree I'm not going to block this.

Reply:
> > Why shouldn't it be part of the type name if the whole purpose of the type is to represent a default value? If we'd only change the parameter name we make this visible only to the function signature, not the type as a whole.
> 
> I think that is here where we disagree. The purpose of the type is to represent the kind of a particular struct. It is in the `InjectPlacementMetadata` that it takes the form of default value. Outside of the function scope, it is not a default value. You said that it represents a default value, but, thinking outside of the `InjectPlacementMetadata`, the default value of what? IMO, it becomes a default value when used inside the `InjectPlacementMetadata`, and only there; for this reason to me it is better to change the parameter name.

This enum is strictly for the purpose of defining a default node placement, hence its name. It does not, and should not, have any other meaning. Outside of the `InjectPlacementMetadata` function it could be used but only for this specific meaning.

You can't just refer to the values themselves without considering their type and then make use of them in a different context. This is why golang supports type branding (i.e. creating new type names out of existing types, like I did here for `int`) in the first place.

In Kubernetes, for example, the value `"Exists"` is defined as part of different types:
```go
TolerationOpExists TolerationOperator = "Exists"
```
```go
ScopeSelectorOpExists       ScopeSelectorOperator = "Exists"
```
```go
NodeSelectorOpExists       NodeSelectorOperator = "Exists"
```

I think that following your reasoning, they could have just used a value for any purpose. But the reason for doing so is to give a value a very specific context in which it is relevant. You can't use `TolerationOpExists` instead of `NodeSelectorOpExists` just because the underlying value is the same. In the same way, you can't just use `AnyNode` to represent anything you want - it is specifically dedicated for default node placement values.

[1] https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/pkg/apis/core/types.go#L2934
[2] https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/pkg/apis/core/types.go#L3223
[3] https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/pkg/apis/core/types.go#L5780

> The same will happen for every structs. We have vms, pods, etc... These types can be used as default for something in some circumstances. But we will change the parameter name, not the type because a pod will always be a pod. This is my reasoning :)

The purpose of vms and pods is not to serve as default values. Their usage is wide. This is not true for the type I'm introducing.

> > But if there's a hypothetical need to change the name of a completely internal type, are you trying to imply that it's a bad thing to do? Can you explain why or what your concerns are? IMHO doing so is completely valid and again I'm not sure exactly what's the concern / risk / problem and the reasoning for it.
> 
> Again, a type should describe the type(kind or whatever), not the behavior :)
> 
> Maybe we just have different opinions about it.

If that's true, can you give me a reason for doing the following?

```go
type MyCoolType int
```

The type is int. Why would we want to re-brand it with a different name if not to describe the usage of this type for a specific context?

Reply:
> > Can you explain? How would the signature reflect how the parameter will be used without the default?
> 
> `DefaultInfraComponentsNodePlacementOption` vs `InfraComponentsNodePlacementOption`, I just understand a param of this type will be used for the node placement of infra components, `Default` in the type doesn't convey anything, what should default mean? I'm passing an explicit value, how does `Default` tell me how the function will behave? My point is, it's not clear one way or the other.

The default emphasizes that these values are relevant specifically in the context of providing a default node placement. So it lets you understand something you wouldn't have understood otherwise, and even worse, you might end up using it for a different purpose which its not designed for.

> We're full of examples in our APIs where default values are documented with comments and this is because there's no such thing as a default type in a statically compiled language there can only be default values, but those depend on how the function is used and the value semantics. A default type by itself is a contradiction.

It's not a default type, it a type of default values. I'm not sure what you mean that there is "no such thing" in compiled languages. Go supports type branding specifically to re-brand the same type but give it a concrete context, and that's what I'm doing here. Go then lets you request a type of that specific re-branded type, so it's not operating solely on the basis of value semantics. Further to that, I previously showed that Kubernetes uses this pattern for about 60 times in their code base. I'm not sure why you're picturing it as contradictory or as something that's obviously wrong.

> I think what you want to mean is that this value will be used by default if no explicit `componentConfig` is passed in, but the `Default` in the type doesn't tell me that, when I read a type name it must tell me something about the type itself not about how some function using that type should behave when another param is nil.

It's not about how "some function is using it" - it is the sole purpose of this type and why it exists. The fact that it's being used by a certain function is an implementation detail, but nevertheless the type should inform you that what it represents is a set of default values for node placement and shouldn't be used for anything else, no matter in which function.

> I think the function is ill-defined, why pass both `componentConfig` and `nodePlacementOption` explicitly? They refer to the same thing, there's a cross dependency that is not clear whether you put the `Default` in the type name or not. One of the issue is that this function is both used for deployments and daemonsets, and this is why you need the additional parameter, otherwise you wouldn't, you could just encode in the function that if `componentConfig` is nill then the default behavior is `X` or `Y`.

Right. If it was for deployments I wouldn't have needed it.
But the reason for passing both `componentConfig` and `nodePlacementOption` is because this function is used for many different objects. For most objects you'd want a way of declaring how to default node placement (i.e., as default means, only if it's not already specified by the `componentConfig`)

> Anyway as I said, even if I don't agree I'm not going to block this.

:+1:

Reply:
As I said, I think we just have different opinions :)
For sure, nothing that will block this PR from getting merged.

Reply:
Thanks @acardace and @fossedihelm for the discussion and for the elaborations :heart: